staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go

```diff
@@ -269,15 +291,20 @@ func (s *Validator) validateMap(ctx context.Context, fldPath *field.Path, sts *s
 	return errs, remainingBudget
 }
 
-func (s *Validator) validateArray(ctx context.Context, fldPath *field.Path, sts *schema.Structural, obj []interface{}, costBudget int64) (errs field.ErrorList, remainingBudget int64) {
+func (s *Validator) validateArray(ctx context.Context, fldPath *field.Path, sts *schema.Structural, obj, oldObj []interface{}, costBudget int64) (errs field.ErrorList, remainingBudget int64) {
 	remainingBudget = costBudget
 	if remainingBudget < 0 {
 		return errs, remainingBudget
 	}
+
+	// only map-type lists support self-oldSelf correlation for cel rules. if this isn't a
+	// map-type list, then makeMapList returns an implementation that always returns nil
+	correlatableOldItems := makeMapList(sts, makeKeyStrategy(sts), oldObj)
```

Comment:
change the signature to `makeMapList(sts, oldObj)` and let `makeMapList` construct the key strategy internally?

Reply:
This was the original signature, but I changed it in order to write tests that cover key collisions. I'll change it back as part of implementing your other suggestion that would eliminate collisions altogether.